library/alloc/src/vec/in_place_drop.rs

```diff
@@ -23,17 +27,24 @@ impl<T> Drop for InPlaceDrop<T> {
     }
 }
 
-// A helper struct for in-place collection that drops the destination allocation and elements,
-// to avoid leaking them if some other destructor panics.
-pub(super) struct InPlaceDstBufDrop<T> {
-    pub(super) ptr: *mut T,
+// A helper struct for in-place collection that drops the destination items together with
+// the source allocation - i.e. before the reallocation happened - to avoid leaking them
+// if some other destructor panics.
+pub(super) struct InPlaceDstDataSrcBufDrop<Src, Dest> {
+    pub(super) ptr: *mut Dest,
     pub(super) len: usize,
-    pub(super) cap: usize,
+    pub(super) src_cap: usize,
+    pub(super) src: PhantomData<Src>,
 }
 
-impl<T> Drop for InPlaceDstBufDrop<T> {
+impl<Src, Dest> Drop for InPlaceDstDataSrcBufDrop<Src, Dest> {
     #[inline]
     fn drop(&mut self) {
-        unsafe { super::Vec::from_raw_parts(self.ptr, self.len, self.cap) };
+        let _drop_allocation = unsafe {
+            RawVec::<Src>::from_raw_parts_in(self.ptr.cast::<Src>(), self.src_cap, Global)
+        };
+        unsafe {
```

Comment:
Any reason not to merge these `unsafe` blocks?

Reply:
The invisible SAFETY comments on top of each block of course. 😛

(Idk why they're not required in `alloc`)

I can merge them.